src/machines/SyncMachine.ts

```diff
@@ -0,0 +1,298 @@
+              actions: ['storeEvent'],
+              cond: 'vertexAccepted',
+              target: CONNECTED_STATES.handlingVertexAccepted,
+            }, {
+              actions: ['storeEvent'],
+              target: CONNECTED_STATES.handlingUnhandledEvent,
+            }],
+          },
+        },
+        [CONNECTED_STATES.handlingUnhandledEvent]: {
+          id: CONNECTED_STATES.handlingUnhandledEvent,
+          invoke: {
+            src: 'updateLastSyncedEvent',
+            onDone: {
+              actions: ['sendAck'],
+              target: 'idle',
+            },
+            onError: `#${SYNC_MACHINE_STATES.ERROR}`,
+          },
+        },
+        [CONNECTED_STATES.handlingMetadataChanged]: {
+          id: 'handlingMetadataChanged',
+          initial: 'detectingDiff',
+          states: {
+            detectingDiff: {
+              invoke: {
+                src: 'metadataDiff',
+                onDone: { actions: ['metadataDecided'] },
+              },
+              on: {
+                METADATA_DECIDED: [
+                  { target: `#${CONNECTED_STATES.handlingVoidedTx}`, cond: 'metadataVoided', actions: ['unwrapEvent'] },
+                  { target: `#${CONNECTED_STATES.handlingUnvoidedTx}`, cond: 'metadataUnvoided', actions: ['unwrapEvent'] },
+                  { target: `#${CONNECTED_STATES.handlingVertexAccepted}`, cond: 'metadataNewTx', actions: ['unwrapEvent'] },
+                  { target: `#${CONNECTED_STATES.handlingFirstBlock}`, cond: 'metadataFirstBlock', actions: ['unwrapEvent'] },
+                  { target: `#${CONNECTED_STATES.handlingUnhandledEvent}`, cond: 'metadataIgnore' },
+                ],
+              },
+            },
+          },
+        },
+        // We have the unchanged guard, so it's guaranteed that this is a new tx
+        [CONNECTED_STATES.handlingVertexAccepted]: {
+          id: CONNECTED_STATES.handlingVertexAccepted,
+          invoke: {
+            src: 'handleVertexAccepted',
+            data: (_context: Context, event: Event) => event,
+            onDone: {
+              target: 'idle',
+              actions: ['sendAck', 'storeEvent'],
+            },
+            onError: `#${SYNC_MACHINE_STATES.ERROR}`,
+          },
+        },
+        [CONNECTED_STATES.handlingVoidedTx]: {
+          id: CONNECTED_STATES.handlingVoidedTx,
+          invoke: {
+            src: 'handleVoidedTx',
+            data: (_context: Context, event: Event) => event,
+            onDone: {
+              target: 'idle',
+              actions: ['storeEvent', 'sendAck'],
+            },
+            onError: `#${SYNC_MACHINE_STATES.ERROR}`,
+          },
+        },
+        [CONNECTED_STATES.handlingUnvoidedTx]: {
+          id: CONNECTED_STATES.handlingUnvoidedTx,
+          invoke: {
+            src: 'handleUnvoidedTx',
+            data: (_context: Context, event: Event) => event,
+            onDone: {
+              target: `#${CONNECTED_STATES.handlingVertexAccepted}`,
```

Comment:
Why do we add the `#` on the state? couldn't find anything about this on the docs

Reply:
It's an [identified state node](https://xstate.js.org/docs/guides/ids.html#identifying-state-nodes), this way it's easier to transition